influxdb3_write/src/last_cache.rs

```diff
@@ -0,0 +1,498 @@
+use std::{any::Any, collections::VecDeque, sync::Arc};
+
+use arrow::{
+    array::{
+        ArrayRef, BooleanBuilder, Float64Builder, GenericByteDictionaryBuilder, Int64Builder,
+        RecordBatch, StringBuilder, StringDictionaryBuilder, TimestampNanosecondBuilder,
+        UInt64Builder,
+    },
+    datatypes::{DataType, GenericStringType, Int32Type, SchemaRef, TimeUnit},
+    error::ArrowError,
+};
+use async_trait::async_trait;
+use datafusion::{
+    datasource::{TableProvider, TableType},
+    execution::context::SessionState,
+    logical_expr::{Expr, TableProviderFilterPushDown},
+    physical_plan::{memory::MemoryExec, ExecutionPlan},
+};
+use hashbrown::{HashMap, HashSet};
+use indexmap::IndexMap;
+use iox_time::Time;
+use parking_lot::RwLock;
+use schema::TIME_COLUMN_NAME;
+
+use crate::{
+    catalog::LastCacheSize,
+    write_buffer::{buffer_segment::WriteBatch, FieldData, Row},
+};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("invalid cache size")]
+    InvalidCacheSize,
+    #[error("last cache already exists for database and table")]
+    CacheAlreadyExists,
+}
+
+/// A two level hashmap storing Database Name -> Table Name -> LastCache
+///
+/// There are two lock levels, one at the top and one at the bottom:
+/// - Top: lock the entire cache for creating new entries
+/// - Bottom: lock an individual cache for pushing in new data
+type CacheMap = RwLock<HashMap<String, HashMap<String, RwLock<LastCache>>>>;
```

Comment:
I think you'd be better off just having a single RwLock around this whole struct. New last caches are only added when a new table or database is created. When writes come in, they're always for multiple databases & tables, so you end up grabbing a bunch of individual locks. Also, the write lock is only obtained on the flush interval for the WAL (which for now is 10ms, but it's going to increase in a later refactoring I have planned).

You end up having to do a bunch more locking for writes and for reads with the double lock that won't really pay off.

Reply:
I changed it so that there is only one top-level lock, and now it is a three-level hashmap, so that it also stores multiple caches per table.